crates/wasi-http/src/p3/request.rs

```diff
@@ -348,3 +494,126 @@ pub async fn default_send_request(
+        fn http(&mut self) -> WasiHttpCtxView<'_> {
+            WasiHttpCtxView {
+                ctx: &mut self.http,
+                table: &mut self.table,
+            }
+        }
+    }
+
+    #[tokio::test]
+    async fn test_request_into_http_schemes() -> Result<()> {
+        let schemes = vec![Some(Scheme::HTTP), Some(Scheme::HTTPS), None];
+        let engine = Engine::default();
+
+        for scheme in schemes {
+            let (req, fut) = Request::new(
+                Method::GET,
+                scheme.clone(),
+                Some(Authority::from_static("example.com")),
+                Some(PathAndQuery::from_static("/path?query=1")),
+                HeaderMap::new(),
+                None,
+                Full::new(Bytes::from_static(b"body"))
+                    .map_err(|x| match x {})
+                    .boxed(),
+            );
+            let mut store = Store::new(&engine, TestCtx::new());
+            let http_req = req.into_http(&mut store, async { Ok(()) }).unwrap();
+            assert_eq!(http_req.method(), Method::GET);
+            let expected_scheme = scheme.unwrap_or(Scheme::HTTPS); // default scheme
+            assert_eq!(
+                http_req.uri(),
+                &http::Uri::from_str(&format!(
+                    "{}://example.com/path?query=1",
+                    expected_scheme.as_str()
+                ))
+                .unwrap()
+            );
+            let body_bytes = http_req.into_body().collect().await?;
+            assert_eq!(*body_bytes.to_bytes(), *b"body");
+            let mut cx = Context::from_waker(Waker::noop());
+            let mut fut = Box::pin(fut);
+            let result = fut.as_mut().poll(&mut cx);
+            assert!(matches!(result, futures::task::Poll::Ready(Ok(()))));
+        }
+
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_request_into_http_invalid_content_length() -> Result<()> {
+        let engine = Engine::default();
+        let mut headers = HeaderMap::new();
+        headers.insert(
+            http::header::CONTENT_LENGTH,
+            HeaderValue::from_static("invalid"),
+        );
+
+        let (req, fut) = Request::new(
+            Method::GET,
+            Some(Scheme::HTTP),
+            Some(Authority::from_static("example.com")),
+            None,
+            headers,
+            None,
+            Full::new(Bytes::new()).map_err(|x| match x {}).boxed(),
+        );
+        let mut store = Store::new(&engine, TestCtx::new());
+        let result = req.into_http(&mut store, async {
+            Err(ErrorCode::InternalError(Some("uh oh".to_string())))
+        });
+        assert!(matches!(result, Err(_)));
+
+        let mut cx = Context::from_waker(Waker::noop());
+        let result = Box::pin(fut).as_mut().poll(&mut cx);
+        // `fut` returns Ok(()), as the error in `into_http` currently drops the body before anything is sent
+        // is this desired behavior?
```

Comment:
If I understand correctly, `into_http` fails before it converts the body, which causes this - that makes sense to me, but I think converting the body should be one of the very first things the function should do, which would maybe fix the issue? (see also comments above as to why the order matters)